### PR TITLE
fix: updated datatypes references

### PR DESCRIPTION
**Pull request by ABHISHEK-821005**
if all the price are same then we can do it in o(n).
improved time complexity.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [ ] Relevant documentation/comments is changed or added
- [ ] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [ ] Search previous suggestions before making a new one, as yours may be a duplicate.
- [ ] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1388"></a>

### CODE REVIEW

File: dynamic_programming/cut_rod.cpp [hunk 1]
```diff
@@ -68,17 +68,71 @@ int maxProfitByCuttingRod(const std::array<int, T> &price, const int n) {
  * @brief Function to test above algorithm
  * @returns void
  */
+
+
+
+template <size_t T>
+bool WhatIfAllPricesAreSame(const std::array<int, T> &price, const int n){
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:75]: This function should be inside the namespaces above.
It is also recommended to use `uint64_t` in the place of `int`.
```suggestion
bool whatIfAllPricesAreSame(const std::array<int, T> &price, const uint64_t &n){
```
This function is also missing proper documentation (see #962 as a reference).

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: dynamic_programming/cut_rod.cpp (hunk 1 above)
Comment (ABHISHEK-821005) [reply dynamic_programming/cut_rod.cpp]: i have done changes as requested, please take a look.

File: dynamic_programming/cut_rod.cpp [hunk 2]
```diff
@@ -61,24 +61,82 @@ int maxProfitByCuttingRod(const std::array<int, T> &price, const int n) {
...
+    std::array<int, n1> price1 = {1, 1,1,1,1,1,1,1};  // price array
     const int max_profit1 =
         dynamic_programming::cut_rod::maxProfitByCuttingRod(price1, n1);
     const int expected_max_profit1 = 22;
+
+    if(dynamic_programming::cut_rod::WhatIfAllPricesAreSame(price1,n1)){
+        std::cout << "Maximum profit with " << n1 << " inch road is " <<(n1)*price1[0]
+              << std::endl;
+    }
+    else{
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:130]:
```suggestion
    else {
```

File: dynamic_programming/cut_rod.cpp [hunk 3]
```diff
@@ -61,24 +61,82 @@ int maxProfitByCuttingRod(const std::array<int, T> &price, const int n) {
...
+
     assert(max_profit1 == expected_max_profit1);
     std::cout << "Maximum profit with " << n1 << " inch road is " << max_profit1
               << std::endl;
 
+    }
+
+
+
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:139]:
```suggestion
```

File: dynamic_programming/cut_rod.cpp [hunk 4]
```diff
@@ -61,24 +61,82 @@ int maxProfitByCuttingRod(const std::array<int, T> &price, const int n) {
...
 }  // namespace dynamic_programming
 
 /**
  * @brief Function to test above algorithm
  * @returns void
  */
+
+
+
+
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:117]:
```suggestion
```

File: dynamic_programming/cut_rod.cpp [hunk 5]
```diff
@@ -61,24 +61,82 @@ int maxProfitByCuttingRod(const std::array<int, T> &price, const int n) {
     delete[] profit;
     return ans;  // returning maximum profit
 }
+
+
+
+template <size_t T>
+bool WhatIfAllPricesAreSame(const std::array<int, T> &price, const uint64_t &n){
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:68]: Missing function documentation.

File: dynamic_programming/cut_rod.cpp [hunk 6]
```diff
@@ -61,24 +61,82 @@ int maxProfitByCuttingRod(const std::array<int, T> &price, const int n) {
...
 static void test() {
     // Test 1
     const int n1 = 8;                                           // size of rod
-    std::array<int, n1> price1 = {1, 5, 8, 9, 10, 17, 17, 20};  // price array
+    std::array<int, n1> price1 = {1, 1,1,1,1,1,1,1};  // price array
     const int max_profit1 =
         dynamic_programming::cut_rod::maxProfitByCuttingRod(price1, n1);
     const int expected_max_profit1 = 22;
+
+    if(dynamic_programming::cut_rod::WhatIfAllPricesAreSame(price1,n1)){
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:126]:
```suggestion
    if (dynamic_programming::cut_rod::WhatIfAllPricesAreSame(price1,n1)) {
```

File: dynamic_programming/cut_rod.cpp (hunk 5 above)
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:68]:
```suggestion
bool WhatIfAllPricesAreSame(const std::array<int32_t, T> &price, const int32_t &n){
```

File: dynamic_programming/cut_rod.cpp [hunk 7]
```diff
@@ -61,24 +61,82 @@ int maxProfitByCuttingRod(const std::array<int, T> &price, const int n) {
...
+ /
+CR(0)
+
+
+   if every length has same price , you would definitely want the rod of length 1, with n quantities.
+    which will give us maximum profits and maximum cuts.
+
+    */
+
+    const int temp=price[0];
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:97]:
```suggestion
    const int16_t temp=price[0];
```

File: dynamic_programming/cut_rod.cpp [hunk 8]
```diff
@@ -61,24 +61,82 @@ int maxProfitByCuttingRod(const std::array<int, T> &price, const int n) {
...
  * @brief Function to test above algorithm
  * @returns void
  */
+
+
+
+
 static void test() {
     // Test 1
     const int n1 = 8;                                           // size of rod
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:120]:
```suggestion
    const int16_t n1 = 8;                                           // size of rod
```

File: dynamic_programming/cut_rod.cpp [hunk 9]
```diff
@@ -61,24 +61,82 @@ int maxProfitByCuttingRod(const std::array<int, T> &price, const int n) {
...
-    std::array<int, n1> price1 = {1, 5, 8, 9, 10, 17, 17, 20};  // price array
+    std::array<int, n1> price1 = {1, 1,1,1,1,1,1,1};  // price array
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:121]:
```suggestion
    std::array<int32_t, n1> price1 = {1, 1,1,1,1,1,1,1};  // price array
```

File: dynamic_programming/cut_rod.cpp [hunk 10]
```diff
@@ -61,24 +61,82 @@ int maxProfitByCuttingRod(const std::array<int, T> &price, const int n) {
...
     const int max_profit1 =
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:122]:
```suggestion
    const int64_t max_profit1 =
```

File: dynamic_programming/cut_rod.cpp [hunk 11]
```diff
@@ -61,24 +61,82 @@ int maxProfitByCuttingRod(const std::array<int, T> &price, const int n) {
...
         dynamic_programming::cut_rod::maxProfitByCuttingRod(price1, n1);
     const int expected_max_profit1 = 22;
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:124]:
```suggestion
    const int32_t expected_max_profit1 = 22;
```

File: dynamic_programming/cut_rod.cpp [hunk 12]
```diff
@@ -61,24 +61,82 @@ int maxProfitByCuttingRod(const std::array<int, T> &price, const int n) {
     delete[] profit;
     return ans;  // returning maximum profit
 }
+
+
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:65]:
```suggestion
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
On function `maxProfitByCuttingRod`, change to `int32_t` on the `price` parameter, change to `int32_t` on the `n` parameter and pass it by reference.

**Comment by ABHISHEK-821005** [comment]
@Panquesito7  please merge this commit if approved or suggest some changes

File: dynamic_programming/cut_rod.cpp [hunk 13]
```diff
@@ -70,25 +70,26 @@ int maxProfitByCuttingRod(const std::array<int, T> &price, const int n) {
...
-    std::array<int, n1> price1 = {1, 5, 8, 9, 10, 17, 17, 20};  // price array
-    const int max_profit1 =
+    const int16_t n1 = 8;                                        // size of rod
+   std::array<int32_t, n1> price1 = {1,2,4,6,8,45,21,9};  // price array
+   const int64_t max_profit1 =
         dynamic_programming::cut_rod::maxProfitByCuttingRod(price1, n1);
-    const int expected_max_profit1 = 22;
-    assert(max_profit1 == expected_max_profit1);
+     const int64_t expected_max_profit1 = 47;
+     assert(max_profit1 == expected_max_profit1);
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:78]:
```suggestion
    const int64_t expected_max_profit1 = 47;
    assert(max_profit1 == expected_max_profit1);
```

File: dynamic_programming/cut_rod.cpp [hunk 14]
```diff
@@ -70,25 +70,26 @@ int maxProfitByCuttingRod(const std::array<int, T> &price, const int n) {
...
         31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
         41, 42, 43, 44, 45, 46, 47, 48, 49, 50};
-    const int max_profit2 =
+
+      const int64_t max_profit2=
         dynamic_programming::cut_rod::maxProfitByCuttingRod(price2, n2);
-    const int expected_max_profit2 = 90;
-    assert(max_profit2 == expected_max_profit2);
+         const int32_t expected_max_profit2 = 90;
+       assert(max_profit2 == expected_max_profit2);
```
Comment (Panquesito7) [inline dynamic_programming/cut_rod.cpp:92]:
```suggestion
    const int32_t expected_max_profit2 = 90;
    assert(max_profit2 == expected_max_profit2);
```

**Comment by ABHISHEK-821005** [comment]
@Panquesito7 please take a look now, i have changed the code as you said.
@Panquesito7 please merge this commit if approved or suggest some changes ,if any.

**Review by Panquesito7 — APPROVED** [review]
LGTM; good work. 👍

**Comment by ABHISHEK-821005** [comment]
🥰🥰🥰
Thank you so much.

**Comment by ABHISHEK-821005** [comment]
@ayaankhan98 please merge this commit if approved or suggest any changes.

**Comment by ABHISHEK-821005** [comment]
@kvedala please merge this commit if approved or suggest any changes.

**Review by ayaankhan98 — APPROVED** [review]

**Pull request merged**